### PR TITLE
[test] Add test for listaddressgroupings

### PR DESCRIPTION
**Pull request by jimmysong**
Test added as part of wallet-accounts.py. Only the most basic grouping is tested.
Made the file flake8 compliant. Most of the diff is whitespace.

### CODE REVIEW

File: test/functional/wallet-accounts.py [hunk 1]
```diff
@@ -27,74 +29,74 @@ def __init__(self):
         self.node_args = [[]]
 
     def setup_network(self):
-        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir, self.node_args)
+        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir,
+                                 self.node_args)
```
Comment (MarcoFalke) [inline test/functional/wallet-accounts.py]: nit: As mentioned [here](https://github.com/bitcoin/bitcoin/pull/10224#discussion_r112836531), I don't think it makes sense to blindly apply pep8 to all python code. Also, changing white space in unrelated commits makes them harder to cherry-pick to other branches.

Comment (jimmysong) [inline test/functional/wallet-accounts.py]: understood, reverted most of the whitespace diff.

**Comment by jimmysong** [comment]
Reverted some of the whitespace diff.

**Comment by TheBlueMatt** [comment]
Hmm, it would be nice to get some more thorough testing of this in. Useful to have tests that "just call it", but to really qualify it as "tests listaddressgroupings" you might want to also create a few test addresses, create some raw transactions to them, test that addresses shared in the input are grouped which weren't previously, and then test that the groupings are transitive (ie if you have a transaction with inputs from address A & B, and another from addresses B & C, there is now one group with all three grouped, despite it being in separate transactions). If you dont want to bite that much off, still good to have these tests, but maybe dont add it to the list of "tested RPCs" at the top, and maybe we should have a blacklist on the coverage meter at the end.

**Comment by jimmysong** [comment]
@TheBlueMatt Thanks for the suggestion. I'm going to have to study a bit more on exactly how addresses group together in the `wallet-accounts.py` test to get a handle on what the groups should be and shouldn't be. I'm fine with removing that this RPC is tested for now or making this task a bigger one with a more thorough test of listaddressgroupings.

**Comment by jnewbery** [comment]
@jimmysong - I haven't looked at the complete test, so I don't know if this breaks things further down, but I imagine you can do this by:

- generate 102 blocks instead of 101. That means the node has 2 coinbase rewards of 50BTC, which are unlinkable and so should be in different groupings.
- construct a single transaction which sends 25 BTC from each address to an external address. The two originating addresses are now linkable and so should be in the same grouping.

If you make this change, there shouldn't be a change in the balance, so :crossed_fingers: it shouldn't break anything further down in the test.

**Comment by jimmysong** [comment]
@jnewbery @TheBlueMatt I've changed the test per jnewbery's suggestion. Hopefully, this is enough to test? Note it was kind of tricky because of fees.

File: test/functional/wallet-accounts.py [hunk 2]
```diff
@@ -30,16 +31,61 @@ def setup_network(self):
...
+            assert_equal(len(address_group), 1)
+            assert_equal(len(address_group[0]), 2)
+            assert_equal(address_group[0][1], 50)
+            linked_addresses.append(address_group[0][0])
+        linked_addresses.sort()
+
+        # send 50 from each address to a third address not in this wallet
+        # fee is 0.0000612, which is why we need to spend 99.9999388
+        common_address = "msf4WtN1YQKXvNtvdFYt9JBnUD2FB41kjr"
+        node.sendfrom("", common_address, 99.9999388)
```
Comment (MarcoFalke) [inline test/functional/wallet-accounts.py]: You might want to try if subtractfromfee helps here.

```py
node.sendmany(fromaccount="",amounts={ common_address:100},subtractfeefrom=[common_address], minconf=1)

Comment (jimmysong) [inline test/functional/wallet-accounts.py]: Thanks! That works great!

**Comment by jimmysong** [comment]
changed so that there's no need to hard-code fees per @MarcoFalke's suggestion.

**Comment by jnewbery** [comment]
Looks good @jimmysong . Thanks for persisting with this.

I would personally prefer to do the second `listaddressgroupings` before you generate a new block, so the only differences between the first and second calls are:
- in the first call, the addresses are in two separate groupings and in the second call they're in the same grouping
- in the first call, both addresses have balance 50 and in the second call both addresses have balance 0.

I have an example here that does that: https://github.com/jnewbery/bitcoin/tree/pr10255 . Feel free to take it for this PR.

Another suggestion: you may be able to remove the fee entirely by using the {create,fund,send}rawtransaction RPCs instead of sendtomany. I'm not sure if that makes the test more straightforward. Up to you if you want to try that out.

**Comment by jimmysong** [comment]
Thanks @jnewbery. I've changed per your suggestions. Of course, I didn't see your tree until I had coded it up, but it's almost exactly the same. Also, rebased and squashed.

**Comment by jnewbery** [comment]
Looks great. utACK dadfee38e8fdf2b384d0991253865ebf9fbdb07f

**Comment by MarcoFalke** [comment]
utACK dadfee38e8fdf2b384d0991253865ebf9fbdb07f
